.buildkite/generator.jl

```diff
@@ -0,0 +1,229 @@
+        PROJECT = dirname(PROJECT)
+    end
+
+    if isempty(PROJECT)
+        println(" --> Skipping as we could not find a build_tarballs.jl")
+        continue
+    end
+
+    # Ignore RootFS stuff, we'll do that manually
+    if startswith(PROJECT, "0_RootFS/")
+        println(" --> Skipping as it's within 0_RootFS/")
+        continue
+    end
+
+    NAME = basename(PROJECT)
+    # Ignore stuff in our excluded projects
+    if NAME in EXCLUDED_NAMES
+        println(" --> Skipping as it's excluded")
+        continue
+    end
+
+    # Otherwise, emit a build with `PROJECT` set to `${PROJECT}`
+    println(" --> Accepted!")
+    push!(PROJECTS_ACCEPTED, PROJECT)
+end
+
+if isempty(PROJECTS_ACCEPTED)
+    annotate("No projects accepted", style="success")
+    exit()
+end
+
+YGGDRASIL_BASE = dirname(@__DIR__)
+julia(args) = `$(Base.julia_cmd()) --project=$(YGGDRASIL_BASE)/.ci $args`
+
+# Next, we're going to ensure that our BB is up to date and precompiled
+julia(`-e "import Pkg; Pkg.instantiate(); Pkg.precompile()"`) |> exec
+
+# We're going to snarf out the BB and BBB tree hashes and combine them to be used later in our build cache
+BB_HASH = julia(```
+    -e "using Pkg, SHA; \
+        gethash(uuid) = collect(Pkg.Types.Context().env.manifest[Pkg.Types.UUID(uuid)].tree_hash.bytes); \
+        print(bytes2hex(sha256(vcat( \
+            gethash(\"7f725544-6523-48cd-82d1-3fa08ff4056e\"), \
+            gethash(\"12aac903-9f7c-5d81-afc2-d9565ea332ae\"), \
+        ))));"
+```) |> cmd -> read(cmd, String)
+
+# Next, for each project, download its sources. We do this by generating meta.json
+# files, then parsing them with `download_sources.jl`
+
+TEMP = mktempdir()
+
+for PROJECT in PROJECTS_ACCEPTED
+    NAME = basename(PROJECT)
+
+    # We always invoke a `build_tarballs.jl` file from its own directory
+    cd(PROJECT) do
+        println("Generating meta.json...")
+        JSON_PATH = "$(TEMP)/$(NAME).meta.json"
+        julia(`--compile=min ./build_tarballs.jl --meta-json="$(JSON_PATH)"`) |> exec
+
+        println("Downloading sources...")
+        julia(`$(YGGDRASIL_BASE)/.ci/download_sources.jl "$(JSON_PATH)" $(TEMP)/$(NAME).platforms.list`) |> exec
+    end
+end
+
+println("Determining builds to queue...")
+for PROJECT in PROJECTS_ACCEPTED
+    NAME = basename(PROJECT)
+
+    # "project source hash" is a combination of meta.json (to absorb
+    # changes from include()'ing a `common.jl`) as well as the entire
+    # tree the project lives in (to absorb changes from patches)
+    # In order to support symlinked directories as dependencies we calculate
+    # the tree hash on a TMP_PROJECT using `cp -RL` to resolve the symlink and
+    # hash the actual content.
+
+    TMP_PROJECT = mktempdir()
+    exec(`cp -RL $(PROJECT) $(TMP_PROJECT)`)
+    TREE_HASH = julia(`-e "using Pkg; print(bytes2hex(Pkg.GitTools.tree_hash(\"$(TMP_PROJECT)\")))"`) |> cmd->read(cmd, String)
```

Comment:
```suggestion
    TREE_HASH = julia(`-e "using Pkg; print(bytes2hex(Pkg.GitTools.tree_hash(\"$(TMP_PROJECT)\")))"`) |> readchomp
```
but can probably avoid shelling out and run the command here?